Comprehensively test supported/unsupported field type:agg combinations

polyfractal opened the pull request:
This adds a test to AggregatorTestCase that allows us to programmatically verify that an aggregator supports or does not support all available field types.  It fetches the list of registered field type parsers, creates a MappedFieldType from the parser and then attempts to run  a basic agg against the field. 

A supplied list of supported VSTypes are then compared against the output (success or exception) and suceeds or fails the test accordingly.

Still a WIP, need to tidy up, add javadocs, add another "main" agg or two (histo, etc).  Putting it up in case folks want to have a look before I spend more time on it.

Note: to make this work, the PR cherry-picks over the `getValuesSourceType()` component of https://github.com/elastic/elasticsearch/pull/51503.  This will be added in eventually from the VS refactor branch, but is technically "unused" code outside of the tests until then.

Note 2: does not support field types added from a module at the moment.

not-napoleon committed:
Put values source types on fields (#51503)

Commit:
Comprehensively test supported/unsupported field type:aggs combinations

This adds a test to AggregatorTestCase that allows us to programmatically
verify that an aggregator supports or does not support a particular
field type.  It fetches the list of registered field type parsers,
creates a MappedFieldType from the parser and then attempts to run
a basic agg against the field.

A supplied list of supported VSTypes are then compared against the
output (success or exception) and suceeds or fails the test accordingly.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Aggregations)

polyfractal commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+        return Collections.emptyList();
+    }
+
+    public void testSupportedFieldTypes() throws IOException {
```
The meat of the PR is here.

polyfractal commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -92,14 +94,6 @@ public RangeFieldType fieldType() {
         }
 
-        @Override
-        public Builder docValues(boolean docValues) {
```
Note: we think this is a bug, I'll open a separate PR fixing this

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+                continue;
+            }
+
+            if (mappedType.getKey().equals(ObjectMapper.CONTENT_TYPE)) {
```
I think you missed removing this check when adding the blacklist?

polyfractal replied:
Ayup

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+                Mapper.Builder builder = mappedType.getValue().parse(fieldName, source, new MockParserContext());
+                mapper = (FieldMapper) builder.build(new BuilderContext(settings, new ContentPath()));
+            } catch (Exception e) {
+                fail();
```
Why not just let the exception happen here? The test will still fail, and the exception is likely to be clearer as to what went wrong than just a call to `fail()`

polyfractal replied:
Leftovers from debugging, this should not be here.  WIP :)

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/support/CoreValuesSourceType.java:
```diff
@@ -224,6 +224,72 @@ public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFa
         }
+    },
+    // TODO: Ordinal Numbering sync with types from master
+    IP{
```
I'm not sure how I feel about pulling the specialized values source types here into Master without the code that uses them.  That seems potentially confusing without context - e.g. We have an IP type now, but VSConfig will still resolve IP fields to Bytes because this isn't wired up.  At a minimum, we should leave a comment in big letters that this is test only until the refactor lands (which hopefully is soon, but still)

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
     }
 
+    protected AggregationBuilder createAggBuilderForTypeTest(MappedFieldType fieldType, String fieldName) {
+        return null;
```
Maybe throw here by default instead of returning `null`?  Otherwise we're going to get an NPE deep in the aggregation code when someone forgets to implement this, and it won't be at all clear why.

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+                            fail("Aggregator [" + aggregationBuilder.getType() + "] should not support field type ["
+                                + fieldType.typeName() + "] but executing against the field did not throw an excetion");
+                        }
+                    } catch (Exception e) {
```
Once we are running these tests against aggregations using the VS Registry, we can say reliably what exception should be thrown.  At that point, we should invert this test - if the VSType is not supported, we should use `expectThrows` to confirm we get the right exception, and otherwise fail on any exception.  Probably can't do that now though since our exceptions are a bit inconsistent in master.

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+        }
+    }
+
+    private static InetAddress randomIp(boolean v4) {
```
We've got copies of this same function in `IpRangeTests` and `IpRangeAggregatorTests`, at least.  Maybe put it up in `ESTestCase` and have all three use the same version?

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+
+            if (fieldType.typeName().equals(RangeType.DOUBLE.typeName())) {
+                start = randomDouble();
+                end = randomDoubleBetween(Math.nextUp((Double)start), Double.MAX_VALUE, false);
```
I'm concerned we might introduce some instability by trying to aggregate over arbitrarily large ranges, which isn't really supported (i.e. it eventually trips the circuit breaker).  I'd suggest just just setting the end value to the next value after the start.  Also, you can use `RangeType.nextUp(Object)` for all cases, it's always well defined.

polyfractal replied:
Ah nice, didn't realize the RangeType's had a `nextUp()` on them :+1:

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +560,163 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+            } else if (fieldType.typeName().equals(RangeType.IP.typeName())) {
+                boolean v4 = randomBoolean();
+                start = randomIp(v4);
+                end = randomIp(v4);
```
Can this generate an invalid range if end sorts before start?

Commit:
Cleanup, javadocs, review comments

polyfractal commented:
Tidied up, added javadocs, and cleaned up the comments that were left. 

For posterity, we also decided to remove the CoreValuesSourceTypes that weren't in use (IP, BOOLEAN, DATE) and return the appropriate type today (BYTES/NUMERIC) so that we didn't have dead code hanging around.

Commit:
Tweak for date_nanos

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +562,184 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+            Mapper.Builder builder = mappedType.getValue().parse(fieldName, source, new MockParserContext());
+            FieldMapper mapper = (FieldMapper) builder.build(new BuilderContext(settings, new ContentPath()));
+
+            MappedFieldType fieldType = mapper.fieldType();
```
If `fieldType.isAggregatable()` returns false, we can probably skip the field here, since it shouldn't have a `ValuesSourceType` if it doesn't support aggregations.  That will save folks adding fields that aren't intended to work with aggregations having to update `writeTestDoc`.

polyfractal replied:
Ah good catch :+1:

not-napoleon commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -524,6 +562,184 @@ protected static IndexReader maybeWrapReaderEs(DirectoryReader reader) throws IO
+                rangeType = RangeType.DOUBLE;
+            } else if (fieldType.typeName().equals(RangeType.FLOAT.typeName())) {
+                start = randomFloat();
+                end = RangeType.FLOAT.nextUp(randomFloat());
```
```suggestion
                end = RangeType.FLOAT.nextUp(start);
```

not-napoleon reviewed (approved):
Looks good, I left two small comments.  Really excited to get this in!

polyfractal committed:
Update test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java

Co-Authored-By: Mark Tozzi <mark.tozzi@gmail.com>

Commit:
Skip fields that are not aggregatable

Commit:
Merge remote-tracking branch 'origin/master' into agg_test_fieldmappers

polyfractal commented:
@elasticmachine run elasticsearch-ci/1

Commit:
Merge remote-tracking branch 'origin/master' into agg_test_fieldmappers
